File: recursion/python/solutions/common_prefix.py
"""Solution for common_prefix."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def common_prefix(lst1, lst2):
    """
    Return the longest common prefix of two lists.

    common_prefix(1 -> 2 -> 3, 1 -> 2 -> 5 -> 6) => 1 -> 2
    common_prefix(1 -> 2, 3 -> 4) => ()
    """
    if is_empty(lst1) or is_empty(lst2):
        return EMPTY_LIST

    first1, rest1 = unprepend(lst1)
    first2, rest2 = unprepend(lst2)

    if first1 == first2:
        return prepend(first1, common_prefix(rest1, rest2))
    else:
        return EMPTY_LIST


if __name__ == '__main__':
    # Test cases
    assert to_python_list(common_prefix(Node.from_list([1, 2, 3]), Node.from_list([1, 2, 5, 6]))) == [1, 2]
    assert to_python_list(common_prefix(Node.from_list([1, 2]), Node.from_list([3, 4]))) == []
    assert to_python_list(common_prefix(EMPTY_LIST, Node.from_list([1, 2]))) == []
    assert to_python_list(common_prefix(Node.from_list([1, 2, 3]), Node.from_list([1, 2, 3]))) == [1, 2, 3]

    print("All tests passed!")

File: recursion/javascript/solutions/commonPrefix.js
/**
 * Solution for commonPrefix.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function commonPrefix(list1, list2) {
  if (isEmpty(list1) || isEmpty(list2)) {
    return EMPTY_LIST;
  }

  let [first1, rest1] = unprepend(list1);
  let [first2, rest2] = unprepend(list2);

  if (first1 === first2) {
    return prepend(first1, commonPrefix(rest1, rest2));
  }
  return EMPTY_LIST;
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(commonPrefix(Node.fromArray([1, 2, 3]), Node.fromArray([1, 2, 5, 6])))) === '[1,2]');
  console.assert(JSON.stringify(toArray(commonPrefix(Node.fromArray([1, 2]), Node.fromArray([3, 4])))) === '[]');
  console.assert(JSON.stringify(toArray(commonPrefix(EMPTY_LIST, Node.fromArray([1, 2])))) === '[]');

  console.log('All tests passed!');
}

module.exports = { commonPrefix };
